agent/Retriever.py

```diff
@@ -0,0 +1,91 @@
+from llama_index import (
+    VectorStoreIndex,
+    SummaryIndex,
+    SimpleDirectoryReader,
+    ServiceContext
+)
+from llama_index.agent import OpenAIAgent
+from llama_index.schema import IndexNode
+from llama_index.tools import QueryEngineTool, ToolMetadata
+from llama_index.llms import OpenAI
+from llama_index.retrievers import RecursiveRetriever
+from llama_index.query_engine import RetrieverQueryEngine
+from llama_index.response_synthesizers import get_response_synthesizer
+
+
+class Retriever:
+    """
+    A class to handle the retrieval and querying of documents using Llama Index services.
+
+    Attributes:
+        llm (OpenAI): An instance of the OpenAI model.
+        service_context (ServiceContext): The context for the service.
+        docs (dict): A dictionary holding the loaded documents.
+        query_engine (RetrieverQueryEngine): The query engine for retrieval tasks.
+    """
+
+    def __init__(self, index_path):
+        """
+        Initialize the Retriever with a specified index path.
+
+        Args:
+            index_path (str): The path to the index directory.
+        """
+        self.llm = OpenAI(temperature=0, model="gpt-4-1106-preview")
+        self.service_context = ServiceContext.from_defaults(llm=self.llm)
```

Comment:
Creating a `ServiceContext` could be a heavyweight operation. Please verify that the context is supposed to be initiated with each `Retriever` instance,
        or if it can be shared amongst them as a single instance, improving efficiency.
```suggestion
        # Assuming that ServiceContext can be created once and reused
        self.service_context = shared_service_context
```